src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenSpanBasedStringConcatTests.cs

```diff
@@ -0,0 +1,5715 @@
+        comp.VerifyIL("C.M", """
+            {
+              // Code size       40 (0x28)
+              .maxstack  2
+              .locals init (char V_0,
+                            char V_1)
+              IL_0000:  ldarg.0
+              IL_0001:  ldfld      "char C.c"
+              IL_0006:  stloc.0
+              IL_0007:  ldloca.s   V_0
+              IL_0009:  newobj     "System.ReadOnlySpan<char>..ctor(in char)"
+              IL_000e:  ldarg.0
+              IL_000f:  call       "ref char C.GetC()"
+              IL_0014:  ldind.u2
+              IL_0015:  stloc.1
+              IL_0016:  ldloca.s   V_1
+              IL_0018:  newobj     "System.ReadOnlySpan<char>..ctor(in char)"
+              IL_001d:  call       "string string.Concat(System.ReadOnlySpan<char>, System.ReadOnlySpan<char>)"
+              IL_0022:  call       "void System.Console.Write(string)"
+              IL_0027:  ret
+            }
+            """);
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/66827")]
+    public void ConcatTwo_ReadOnlySpan_MutateLocal()
+    {
+        var source = """
+            using System;
+
+            var c = new C();
+            Console.WriteLine(c.M());
+
+            class C
+            {
+                public string M()
+                {
+                    var c = 'a';
+                    return c + SneakyLocalChange(ref c);
+                }
+
+                private string SneakyLocalChange(ref char local)
+                {
+                    local = 'b';
+                    return "b";
+                }
+            }
+            """;
+
+        var comp = CompileAndVerify(source, expectedOutput: RuntimeUtilities.IsCoreClr8OrHigherRuntime ? "ab" : null, targetFramework: TargetFramework.Net80, verify: RuntimeUtilities.IsCoreClr8OrHigherRuntime ? default : Verification.Skipped);
+
+        comp.VerifyDiagnostics();
+        comp.VerifyIL("C.M", """
+            {
+              // Code size       31 (0x1f)
+              .maxstack  3
+              .locals init (char V_0, //c
+                            char V_1)
+              IL_0000:  ldc.i4.s   97
+              IL_0002:  stloc.0
+              IL_0003:  ldloc.0
+              IL_0004:  stloc.1
+              IL_0005:  ldloca.s   V_1
+              IL_0007:  newobj     "System.ReadOnlySpan<char>..ctor(in char)"
+              IL_000c:  ldarg.0
+              IL_000d:  ldloca.s   V_0
+              IL_000f:  call       "string C.SneakyLocalChange(ref char)"
+              IL_0014:  call       "System.ReadOnlySpan<char> string.op_Implicit(string)"
+              IL_0019:  call       "string string.Concat(System.ReadOnlySpan<char>, System.ReadOnlySpan<char>)"
+              IL_001e:  ret
+            }
+            """);
+    }
+
+    [Theory]
+    [InlineData(null)]
+    [InlineData((int)SpecialMember.System_String__Concat_2ReadOnlySpans)]
+    public void ConcatTwo_ReadOnlySpan_NullConcatArgument(int? missingUnimportantMember)
```

Comment:
> for scenarios that this PR is targeting

Even though there are no spans in resulting codegen of this test, I would still say that this scenario is related to the PR. In this case we test that "remove constant `null` or empty concatenation argument" optimization comes before we try to rewrite concatenation using span-based approach. The codegen of this test is gonna be improved once https://github.com/dotnet/roslyn/issues/72232 is fixed